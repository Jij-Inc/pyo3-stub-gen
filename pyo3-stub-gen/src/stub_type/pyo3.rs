use crate::stub_type::*;
use ::pyo3::{
    basic::CompareOp,
    pybacked::{PyBackedBytes, PyBackedStr},
    pyclass::boolean_struct::False,
    types::*,
    Bound, Py, PyClass, PyRef, PyRefMut,
};
use maplit::hashset;

impl PyStubType for PyAny {
    fn type_output() -> TypeInfo {
        TypeInfo {
            name: "typing.Any".to_string(),
            import: hashset! { "typing".into() },
        }
    }
}

impl<T: PyStubType> PyStubType for Py<T> {
    fn type_input() -> TypeInfo {
        T::type_input()
    }
    fn type_output() -> TypeInfo {
        T::type_output()
    }
}

impl<T: PyStubType + PyClass> PyStubType for PyRef<'_, T> {
    fn type_input() -> TypeInfo {
        T::type_input()
    }
    fn type_output() -> TypeInfo {
        T::type_output()
    }
}

impl<T: PyStubType + PyClass<Frozen = False>> PyStubType for PyRefMut<'_, T> {
    fn type_input() -> TypeInfo {
        T::type_input()
    }
    fn type_output() -> TypeInfo {
        T::type_output()
    }
}

impl<T: PyStubType> PyStubType for Bound<'_, T> {
    fn type_input() -> TypeInfo {
        T::type_input()
    }
    fn type_output() -> TypeInfo {
        T::type_output()
    }
}

macro_rules! impl_builtin {
    ($ty:ty, $pytype:expr) => {
        impl PyStubType for $ty {
            fn type_output() -> TypeInfo {
                TypeInfo {
                    name: $pytype.to_string(),
                    import: HashSet::new(),
                }
            }
        }
    };
}

impl_builtin!(PyInt, "int");
impl_builtin!(PyFloat, "float");
impl_builtin!(PyComplex, "complex");
impl_builtin!(PyList, "list");
impl_builtin!(PyTuple, "tuple");
impl_builtin!(PySlice, "slice");
impl_builtin!(PyDict, "dict");
impl_builtin!(PySet, "set");
impl_builtin!(PyString, "str");
impl_builtin!(PyBackedStr, "str");
impl_builtin!(PyByteArray, "bytearray");
impl_builtin!(PyBytes, "bytes");
impl_builtin!(PyBackedBytes, "bytes");
impl_builtin!(PyType, "type");
impl_builtin!(CompareOp, "int");

macro_rules! impl_simple {
    ($ty:ty, $mod:expr, $pytype:expr) => {
        impl PyStubType for $ty {
            fn type_output() -> TypeInfo {
                TypeInfo {
                    name: concat!($mod, ".", $pytype).to_string(),
                    import: hashset! { $mod.into() },
                }
            }
        }
    };
}

impl_simple!(PyDate, "datetime", "date");
impl_simple!(PyDateTime, "datetime", "datetime");
impl_simple!(PyDelta, "datetime", "timedelta");
impl_simple!(PyTime, "datetime", "time");
impl_simple!(PyTzInfo, "datetime", "tzinfo");
