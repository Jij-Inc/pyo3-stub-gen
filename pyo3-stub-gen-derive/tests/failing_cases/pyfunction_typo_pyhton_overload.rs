use pyo3::prelude::*;
use pyo3_stub_gen_derive::gen_stub_pyfunction;

#[gen_stub_pyfunction(pyhton_overload = r#"
def example(x: int) -> int: ...
"#)]
#[pyfunction]
fn example(x: i32) -> i32 {
    x
}
