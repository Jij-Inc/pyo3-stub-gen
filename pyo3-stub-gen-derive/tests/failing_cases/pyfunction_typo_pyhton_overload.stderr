error: Unknown parameter: pyhton_overload
 --> tests/failing_cases/pyfunction_typo_pyhton_overload.rs:4:23
  |
4 | #[gen_stub_pyfunction(pyhton_overload = r#"
  |                       ^^^^^^^^^^^^^^^

warning: unused import: `pyo3::prelude::*`
 --> tests/failing_cases/pyfunction_typo_pyhton_overload.rs:1:5
  |
1 | use pyo3::prelude::*;
  |     ^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` (part of `#[warn(unused)]`) on by default

error[E0601]: `main` function not found in crate `$CRATE`
  --> tests/failing_cases/pyfunction_typo_pyhton_overload.rs:10:2
   |
10 | }
   |  ^ consider adding a `main` function to `$DIR/tests/failing_cases/pyfunction_typo_pyhton_overload.rs`
